chef/cookbooks/cinder/recipes/ceph.rb

```diff
@@ -17,51 +17,82 @@
 # Recipe:: ceph
 #
 
-glance_servers = search(:node, "roles:glance-server")
-if glance_servers.length > 0
-  glance_pool = glance_servers[0][:glance][:rbd][:store_pool]
-else   
-  glance_pool = nil
-end
+ceph_clients = {}
+ceph_keyrings = {}
 
 node[:cinder][:volumes].each_with_index do |volume, volid|
-  if volume['backend_driver'] == "rbd"
+  unless volume[:backend_driver] == "rbd"
+    next
+  else
+
+    ceph_conf = volume[:rbd][:config_file]
+    admin_keyring = volume[:rbd][:admin_keyring]
+    if File.exists?(admin_keyring)
+      Chef::Log.info("Using external ceph cluster for cinder, with automatic setup.")
+    else
+      Chef::Log.info("Using external ceph cluster for cinder, with no automatic setup.")
+      return
+    end
 
-    check_ceph = Mixlib::ShellOut.new("ceph -k #{volume['rbd']['admin_keyring']} -c #{volume['rbd']['config_file']} -s | grep -q -e 'HEALTH_[OK|WARN]'")
-    check_ceph.run_command
+    cmd = ["ceph", "-k", admin_keyring, "-c", ceph_conf, "-s"]
+    check_ceph = Mixlib::ShellOut.new(cmd)
 
-    if check_ceph.exitstatus == 0
+    Chef::Log.info("Checking ceph cluster status.")
+    if check_ceph.run_command.stdout.match("(HEALTH_OK|HEALTH_WARN)")
+
+      backend_id = "backend-#{volume[:backend_driver]}-#{volid}"
 
-      backend_id = "backend-#{volume['backend_driver']}-#{volid}"
       cinder_user = volume[:rbd][:user]
       cinder_pool = volume[:rbd][:pool]
 
-      cinder_pools = []
-      cinder_pools << cinder_pool
+      ceph_clients[ceph_conf] = {} unless ceph_clients[ceph_conf]
+      ceph_keyrings[ceph_conf] = admin_keyring unless ceph_keyrings[ceph_conf]
+
+      cinder_pools = (ceph_clients[ceph_conf][cinder_user] || []) << cinder_pool
+      ceph_clients[ceph_conf][cinder_user] = cinder_pools
+
+      ceph_pool cinder_pool do
+        ceph_conf ceph_conf
+        admin_keyring admin_keyring
+      end
+    else
+      Chef::Log.info("Ceph cluster for RADOS backend is not ready yet.")
```

Comment:
You still don't mention the name of the backend: there can be multiple rbd backends and we need to know which one this is about... I think you have that in `volume[:backend_name]` (but please double-check)